File: engine/src/physics/particle/particle-force-generator.ts
import { type IVector2, Vector2 } from "../../math/vector2";
import { type IParticlePhysics } from "./iparticle-physics";
import { clamp } from "../../utils/clamp";

export class ParticleForceGenerator {
    private constructor() {}

    public static weightForce(particle: IParticlePhysics): Vector2 {
        const mass = particle.getMass();

        if (mass === 0) {
            return new Vector2();
        }

        return particle.getGravity().toScaled(mass);
    }

    public static dragForce(particle: IParticlePhysics, constant: number): IVector2 {
        if (particle.getMass() === 0) {
            return new Vector2();
        }

        const velocitySquaredMagnitude = particle.getVelocity().getSquaredMagnitude();

        if (velocitySquaredMagnitude <= 0) {
            return new Vector2();
        }

        const dragDirection = particle.getVelocity().toNormalized().scale(-1);
        const dragMagnitude = constant * velocitySquaredMagnitude;

        return dragDirection.scale(dragMagnitude);
    }

    public static frictionForce(particle: IParticlePhysics, constant: number): IVector2 {
        if (particle.getMass() === 0) {
            return new Vector2();
        }

        const frictionDirection = particle.getVelocity().toNormalized().scale(-1);

        return frictionDirection.scale(constant);
    }

    public static gravitationForce(
        particle: IParticlePhysics, other: IParticlePhysics,
        gravity: number,
        minDistance: number, maxDistance: number,
    ): IVector2 {
        if (particle.getMass() === 0) {
            return new Vector2();
        }

        const distance = other.getPosition().toSubtracted(particle.getPosition());

        const squaredMagnitude = clamp(
            distance.getSquaredMagnitude(),
            minDistance,
            maxDistance,
        );

        const attractionDirection = distance.toNormalized();

        const attractionMagnitude = gravity * (particle.getMass() * other.getMass()) / squaredMagnitude;

        return attractionDirection.scale(attractionMagnitude);
    }

    public static springForce(
        particle: IParticlePhysics, anchor: IVector2,
        constant: number,
        restLength: number,
    ): IVector2;
    public static springForce(
        particle: IParticlePhysics, other: IParticlePhysics,
        constant: number,
        restLength: number,
    ): IVector2;
    public static springForce(
        particle: IParticlePhysics, other: IVector2 | IParticlePhysics,
        constant: number,
        restLength: number,
    ): IVector2 {
        const otherPosition: IVector2 = isVector2(other) ? other : other.getPosition();

        const d = particle.getPosition().toSubtracted(otherPosition);

        const displacement = d.getMagnitude() - restLength;

        const springDirection = d.toNormalized();

        const springMagnitude = -constant * displacement;

        return springDirection.scale(springMagnitude);
    }
}

function isVector2(value: unknown | IVector2): value is IVector2 {
    return value instanceof Vector2;
}

File: crates/kphysics/src/particle_force_generator.rs
use kmath::Vector2;

use crate::Particle;

#[inline]
pub const fn weight(particle: &Particle, k: f64) -> Vector2 {
    Vector2::new(0.0, 9.8 * k * particle.mass)
}

#[inline]
pub fn drag(particle: &Particle, k: f64) -> Vector2 {
    let velocity_squared_magnitude = particle.velocity.squared_magnitude();

    if velocity_squared_magnitude <= 0.0 {
        return Vector2::ZERO;
    }

    let drag_direction = particle.velocity.to_normalized().to_scaled(-1.0);

    let drag_magnitude = k * velocity_squared_magnitude;

    drag_direction.to_scaled(drag_magnitude)
}

#[inline]
pub fn friction(particle: &Particle, k: f64) -> Vector2 {
    let friction_direction = particle.velocity.to_normalized().to_scaled(-1.0);

    let friction_magnitude = k;

    friction_direction.to_scaled(friction_magnitude)
}

#[inline]
pub fn gravitation(
    a: &Particle,
    b: &Particle,
    g: f64,
    min_distance: f64,
    max_distance: f64,
) -> Vector2 {
    let disposition = b.position.to_subtracted(&a.position);

    let squared_distance = disposition
        .squared_magnitude()
        .clamp(min_distance, max_distance);

    let attraction_direction = disposition.to_normalized();

    let attraction_magnitude = g * (a.mass * b.mass) / squared_distance;

    attraction_direction.to_scaled(attraction_magnitude)
}
